src/main/java/com/google/devtools/build/lib/bazel/bzlmod/Discovery.java

```diff
@@ -58,14 +68,38 @@ public static ImmutableMap<ModuleKey, InterimModule> run(
           if (depGraph.containsKey(depSpec.toModuleKey())) {
             continue;
           }
+          predecessors.putIfAbsent(depSpec.toModuleKey(), module.getKey());
           unexpandedSkyKeys.add(
               ModuleFileValue.key(depSpec.toModuleKey(), overrides.get(depSpec.getName())));
         }
       }
       SkyframeLookupResult result = env.getValuesAndExceptions(unexpandedSkyKeys);
       for (SkyKey skyKey : unexpandedSkyKeys) {
         ModuleKey depKey = ((ModuleFileValue.Key) skyKey).getModuleKey();
-        ModuleFileValue moduleFileValue = (ModuleFileValue) result.get(skyKey);
+        ModuleFileValue moduleFileValue;
+        try {
+          moduleFileValue =
+              (ModuleFileValue) result.getOrThrow(skyKey, ExternalDepsException.class);
+        } catch (ExternalDepsException e) {
+          // Trace back a dependency chain to the root module. There can be multiple paths to the
+          // failing module, but any of those is useful for debugging.
+          List<ModuleKey> depChain = new ArrayList<>();
+          depChain.add(depKey);
+          ModuleKey predecessor = depKey;
+          while ((predecessor = predecessors.get(predecessor)) != null) {
+            depChain.add(predecessor);
+          }
+          Collections.reverse(depChain);
+          String depChainString =
+              depChain.stream().map(ModuleKey::toString).collect(joining(" -> "));
+          throw new BazelModuleResolutionFunction.BazelModuleResolutionFunctionException(
+              ExternalDepsException.withCauseAndMessage(
+                  FailureDetails.ExternalDeps.Code.BAD_MODULE,
+                  e,
+                  "in module dependency chain %s",
+                  depChainString),
+              SkyFunctionException.Transience.TRANSIENT);
```

Comment:
We are losing the original transience here since for some reason we can only catch the `ExternalDepsException`, not the wrapping `BazelModuleFileException`. I don't know whether this is me or us holding Skyframe wrong.

Reply:
IIUC this is WAI. the impls of SkyFunctionException are not meant to be caught; they're always wrappers around some other exception type along with some transience information to let Skyframe infer whether to attempt re-evaluation. But re-evaluation can only occur if all Skyframe dependencies are in place. So in our case where we're basically re-throwing an exception, it shouldn't matter what our transience is, since the thrower of the original exception needs to be re-evaled first anyway.

On that note, I think the better idiom here is to have `Discovery` throw an `ExternalDepsException`, and have `BazelModuleResolutionFunction` wrap it.

Reply:
Thanks for the explanation! :-)

I updated the commit to follow that pattern.